core/arch/arm/tee/pta_socket.c

```diff
@@ -0,0 +1,335 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <kernel/static_ta.h>
+#include <optee_msg.h>
+#include <optee_msg_supplicant.h>
+#include <pta_socket.h>
+#include <string.h>
+#include <tee/tee_fs_rpc.h>
+
+static uint32_t get_instance_id(struct tee_ta_session *sess)
+{
+	return sess->ctx->ops->get_instance_id(sess->ctx);
+}
+
+static TEE_Result socket_open(struct tee_ta_session *sess, uint32_t param_types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res;
+	paddr_t pa;
+	uint64_t cookie;
+	void *va;
+	struct optee_msg_param msg_params[4];
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT);
+
+	if (exp_pt != param_types) {
+		DMSG("got param_types 0x%x, expected 0x%x",
+		     param_types, exp_pt);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memset(msg_params, 0, sizeof(msg_params));
+
+	va = tee_fs_rpc_cache_alloc(params[1].memref.size, &pa, &cookie);
+	if (!va)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	msg_params[0].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
+	msg_params[0].u.value.a = OPTEE_MRC_SOCKET_OPEN;
+	msg_params[0].u.value.b = get_instance_id(sess);
+
+	msg_params[1].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
+	msg_params[1].u.value.a = params[0].value.b; /* server port number */
+	msg_params[1].u.value.b = params[2].value.a; /* protocol */
+	msg_params[1].u.value.c = params[0].value.a; /* ip version */
+
+	/* server address */
+	msg_params[2].attr = OPTEE_MSG_ATTR_TYPE_TMEM_INPUT;
+	msg_params[2].u.tmem.buf_ptr = pa;
+	msg_params[2].u.tmem.size = params[1].memref.size;
+	msg_params[2].u.tmem.shm_ref = cookie;
+	memcpy(va, params[1].memref.buffer, params[1].memref.size);
+
+	/* socket handle */
+	msg_params[3].attr = OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT;
+
+	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_SOCKET, 4, msg_params);
+
+	if (res == TEE_SUCCESS)
+		params[3].value.a = msg_params[3].u.value.a;
```

Comment:
I wonder how this could have worked. I'm updating tee-supplicant and protocol description to use `value.a` instead.